### PR TITLE
Add from_iso8601_date() presto function

### PR DESCRIPTION
**Pull request by svm1**
Implementing `from_iso8601_date()` presto function. Adding test cases to validate functionality.

Created new PR due to issues with squashing and rebasing on the original.

Resolves https://github.com/facebookincubator/velox/issues/4737

### CODE REVIEW

**Comment by svm1** [comment]
@aditi-pandit @majetideepak finalized my changes here, please review when you get a chance!

File: velox/functions/prestosql/DateTimeFunctions.h [hunk 1]
```diff
@@ -1213,6 +1213,25 @@ struct CurrentDateFunction {
...
 };
 
+template <typename T>
+struct FromISO8601DateFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Date>& result,
+      const arg_type<Varchar>& isoDateStr) {
+    int len = isoDateStr.size();
```
Comment (aditi-pandit) [inline velox/functions/prestosql/DateTimeFunctions.h:1223]: int len = min(isoDateStr.size(), 10);
result = len < 10 ? DATE()->toDays(isoDateStr) : DATE()->toDays(isoDateStr.substr(0, 10))

I think you can work on the original StringViews without all the copies to std::string and substr on it.

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 2]
```diff
@@ -3324,6 +3324,27 @@ TEST_F(DateTimeFunctionsTest, timeZoneMinute) {
...
 }
 
+TEST_F(DateTimeFunctionsTest, fromISO8601FunctionDate) {
+  const auto fromISODate =
+      [&](const std::optional<std::string>& isoDateString) {
+        return evaluateOnce<int32_t>("from_iso8601_date(c0)", isoDateString);
+      };
+
+  EXPECT_EQ(0, fromISODate("1970-01-01T03:19:58.000"));
+  EXPECT_EQ(18297, fromISODate("2020-02-05T14:27:39.000-07:00"));
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:3334]: Maybe add tests for some more extreme values like 9999-12-31 and 0000-01-01

Comment (svm1) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:3334]: Good point, added.

File: velox/functions/prestosql/DateTimeFunctions.h (hunk 1 above)
Comment (svm1) [reply velox/functions/prestosql/DateTimeFunctions.h]: I'm a fan of ternary exps, though avoided to keep clarity/wasn't sure if favored. Condensed the code as such.

I initially tried working on the StringView but it didn't seem possible (doesn't seem to have methods like substr), so extracting the std::string seemed like the best bet.

File: velox/functions/prestosql/DateTimeFunctions.h [hunk 3]
```diff
@@ -15,6 +15,7 @@
  */
 #pragma once
 
+#include <algorithm>
```
Comment (aditi-pandit) [inline velox/functions/prestosql/DateTimeFunctions.h:18]: Why is this needed ?

File: velox/functions/prestosql/DateTimeFunctions.h [hunk 4]
```diff
@@ -1213,6 +1214,20 @@ struct CurrentDateFunction {
...
 };
 
+template <typename T>
+struct FromISO8601DateFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Date>& result,
+      const arg_type<Varchar>& isoDateStr) {
+    int len = std::min((int)isoDateStr.size(), 10);
```
Comment (aditi-pandit) [inline velox/functions/prestosql/DateTimeFunctions.h:1224]: Could you try : 
result = len < 10 ? DATE()->toDays(isoDateStr) : DATE()->toDays(isoDateStr->getString().substr(0, 10)); ?

and remove the isoDateExtracted variable assignment.

File: velox/functions/prestosql/DateTimeFunctions.h (hunk 3 above)
Comment (svm1) [reply velox/functions/prestosql/DateTimeFunctions.h:18]: The `min()` function used below comes from the standard algorithms library.

File: velox/functions/prestosql/DateTimeFunctions.h (hunk 4 above)
Comment (svm1) [reply velox/functions/prestosql/DateTimeFunctions.h:1224]: Modified as such, removed the variable assignment and condensed the operations.

**Review by aditi-pandit — APPROVED** [review]

**Comment by aditi-pandit** [comment]
@svm1 : Please can you rebase the code and ensure all builds pass.

**Comment by aditi-pandit** [comment]
@Yuhta : Please can you take a look.

**Comment by svm1** [comment]
@Yuhta Can you please review when you get a chance?

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 5]
```diff
@@ -3498,6 +3498,29 @@ TEST_F(DateTimeFunctionsTest, timeZoneMinute) {
...
+  EXPECT_EQ(-18297, fromISODate("1919-11-28T23:59:59.999"));
+  EXPECT_EQ(-18297, fromISODate("1919-11-28T23:59:59.999+08:00"));
+  EXPECT_EQ(-2, fromISODate("1969-12-30T19:00:00.000-05:00"));
+  EXPECT_EQ(-719528, fromISODate("0000-01-01T00:00:00.000+00:00"));
+  EXPECT_EQ(2932896, fromISODate("9999-12-31:23:59.999+00:00"));
```
Comment (Yuhta) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:3513]: Why this one is passing?

Comment (svm1) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:3513]: Sorry, I don't understand - is there a particular test there you are referring to? My general idea was that `from_iso8601_date` should convert an ISO-formatted timestamp string to a Date type, and as the Date type is nothing but a 32-bit integer now, representing number of days since the epoch, this should be a sufficient check of the validity of the function.

Comment (svm1) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:3513]: Hmm, I tested it out and it seems that Java implementation of `from_iso8601_date()` is rejecting any ISO-formatted string:

- generating ISO-formatted string from `to_ISO8601()`
```
presto:default> SELECT to_iso8601(TIMESTAMP '2001-08-22 03:04:05.321 Africa/Cairo');
SELECT to_iso8601(TIMESTAMP '2001-08-22 03:04:05.321 Africa/Cairo');
             _col0             
-------------------------------
 2001-08-22T03:04:05.321+03:00 
(1 row)
```
- passing the above output string into `from_iso8601_date()`
```
presto:default> SELECT from_iso8601_date('2001-08-22T03:04:05.321+03:00');
SELECT from_iso8601_date('2001-08-22T03:04:05.321+03:00');
Query 20231027_232729_00005_bta8n failed: Invalid format: "2001-08-22T03:04:05.321+03:00" is malformed at "T03:04:05.321+03:00"
```

Possible bug?

Comment (Yuhta) [reply velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:3513]: I think java is using ' ' instead 'T' here.  Can you match the behavior in Java?

**Comment by mbasmanova** [comment]
@svm1 I'm closing this PR since the function was added in #8914. That PR allows to parse YYYY-MM-DD strings.

However, it turns out that from_iso8601_date function in Presto supports partial inputs as well: YYYY and YYYY-MM. It would be nice to add support the these.

```
presto:di> select from_iso8601_date('123');
   _col0
------------
 0123-01-01
(1 row)

presto:di> select from_iso8601_date('2020-1');
   _col0
------------
 2020-01-01
(1 row)

presto:di> select from_iso8601_date('2020-11');
   _col0
------------
 2020-11-01
(1 row)

presto:di> select from_iso8601_date('2020-11-3');
   _col0
------------
 2020-11-03
(1 row)

presto:di> select from_iso8601_date('2020-11-30');
   _col0
------------
 2020-11-30
(1 row)
```
